src/gflip/gflip_engine.cpp

```diff
@@ -358,10 +443,134 @@ void gflip_engine::run_evaluation(int dtype)
 
 // ---------------------------------------------------------
 
+void gflip_engine::update_tfidf()
+{
+  {
+    ScopedTimer scopedTimer("Adding docs");
+    for (int i=0; i<laserscan_bow.size(); ++i)
+    {
+      add_doc_stats(i);
+    }
+  }
+
+  {
+    ScopedTimer scopedTimer("Updating tf_idf");
+    compute_idfs();
+    compute_tf_idfs();
+  }
+
+  {
+    ScopedTimer scopedTimer(" tf_idf");
+    for(uint i=0;i<laserscan_bow.size();i++)
+      laserscan_bow[i].norm_wgv = norm_gfp(laserscan_bow[i]);
+  }
+
+}
+
+// ---------------------------------------------------------
+
+
+void gflip_engine::update_tfidf(int nscans)
+{
+  {
+    ScopedTimer scopedTimer("Adding scans");
+    for (int i=laserscan_bow.size()- nscans; i<laserscan_bow.size(); ++i)
+    {
+      add_doc_stats(i);
+    }
+  }
+  {
+    ScopedTimer scopedTimer("Updating tf_idf");
+    compute_idfs();
+    compute_tf_idfs();
+  }
+
+  {
+    ScopedTimer scopedTimer("Normalizing tf_idf");
+    for(uint i=0;i<laserscan_bow.size();i++)
+      laserscan_bow[i].norm_wgv = norm_gfp(laserscan_bow[i]);
+  }
+}
+
+
+
+// ---------------------------------------------------------
+
+void gflip_engine::add_doc_stats(int doc_id)
+{
+
+    std::map<int, std::vector<int> >::const_iterator it = laserscan_bow[doc_id].histogram.begin();
+    for(; it != laserscan_bow[doc_id].histogram.end(); ++it)
+    {
+      int word_id = it->first;
+      tf_idf[word_id].term_count_unnormalized.push_back(it->second.size());
+      tf_idf[word_id].num_words.push_back(laserscan_bow[doc_id].w.size());
+      laserscan_bow[doc_id].word_ref[word_id] = tf_idf[word_id].doc_id.size();
+      tf_idf[word_id].doc_id.push_back(doc_id);
+      tf_idf[word_id].term_count.push_back((double)it->second.size() / (double)laserscan_bow[doc_id].w.size());
+      tf_idf[word_id].word_order.push_back(tf_idf_db_ordercache(it->second));
+      tf_idf[word_id].tf_idf_doc_normed.push_back(-1);
+      tf_idf[word_id].wf_idf_doc_normed.push_back(-1);
+      tf_idf[word_id].ntf_idf_doc_normed.push_back(-1);
+    }
+}
+
+void gflip_engine::compute_tf_idfs()
+{
+  for(int doc_id=0;doc_id<(int)laserscan_bow.size();doc_id++)
+  {
+    std::map<int, int>::const_iterator word_id_iter = laserscan_bow[doc_id].word_ref.begin();
+    double sum=0,sum_wf=0,sum_vss=0;
+    for (; word_id_iter!=laserscan_bow[doc_id].word_ref.end(); ++word_id_iter)
+    {
+      int doc_ref = word_id_iter->second;
+      tf_idf_db & db = tf_idf[word_id_iter->first];
+      double val = db.tf_idf_doc_normed[doc_ref]  = db.term_count[doc_ref] * db.idf;
+      double val_wf = db.wf_idf_doc_normed[doc_ref] =  (1 + log(db.term_count_unnormalized[doc_ref]) ) * db.idf;
+      double val_vss = db.ntf_idf_doc_normed[doc_ref] = alpha_vss + ( (1.0 - alpha_vss) * db.term_count_unnormalized[doc_ref]) / laserscan_bow[doc_id].max_histogram;
```

Comment:
`laserscan_bow[doc_id].max_histogram` doesn't change inside the loop, you can have a local variable